File: 3213-count-subarrays-where-max-element-appears-at-least-k-times/count-subarrays-where-max-element-appears-at-least-k-times.cpp
class Solution {
public:
    long long countSubarrays(vector<int>& nums, int k) {
        long long ans = 0;
        int s=0,e=0, sz = nums.size();
        int maxi = *max_element(nums.begin(), nums.end()), cnt = 0;

        while(e < sz){
            if(maxi == nums[e]) cnt++;
            while(cnt >= k){
                if(nums[s] == maxi) cnt--;
                s++;
                ans += sz - e;
            }
            e++;
        }

        return ans;
    }
};

File: 3213-count-subarrays-where-max-element-appears-at-least-k-times/count-subarrays-where-max-element-appears-at-least-k-times.java
class Solution {
    public long countSubarrays(int[] nums, int k) {
        int sz = nums.length, maxi = nums[0];
        long ans = 0, last = -1;
        Queue<Integer> q = new LinkedList<>();

        for(int i=0; i<sz; i++) maxi = Math.max(maxi, nums[i]);
        
        for(int i=0; i<sz; i++){
            if(nums[i] == maxi) q.add(i);
            if(q.size() == k){
                int front = q.poll();
                ans += 1L * (front - last) * (sz-i);
                last = front;
            }
        }

        return ans;
    }
}